LibOS/shim/src/bookkeep/shim_vma.c

```diff
@@ -37,229 +37,327 @@
-    thread->delayed_bkeep_mmap.addr = addr;
-    thread->delayed_bkeep_mmap.length = size;
-    return addr;
+
+    return (void *) DkVirtualMemoryAlloc(addr, size, 0,
+                                         PAL_PROT_WRITE|PAL_PROT_READ);
 }
 
 #undef system_malloc
-#define system_malloc(size) __vma_malloc(size)
+#define system_malloc __malloc
 
 #define OBJ_TYPE struct shim_vma
 #include <memmgr.h>
 
+/*
+ * "vma_mgr" has no specific lock. "vma_list_lock" must be held when
+ * allocating or freeing any VMAs.
+ */
 static MEM_MGR vma_mgr = NULL;
 
+/*
+ * "vma_list" contains a sorted list of non-overlapping VMAs.
+ * "vma_list_lock" must be held when accessing either the vma_list or any
+ * field of a VMA.
+ */
 DEFINE_LISTP(shim_vma);
 static LISTP_TYPE(shim_vma) vma_list = LISTP_INIT;
 static LOCKTYPE vma_list_lock;
 
-static inline bool test_vma_equal (struct shim_vma * tmp,
-                                  const void * addr, uint64_t length)
-{
-    return tmp->addr == addr && tmp->length == length;
-}
-
-static inline bool test_vma_contain (struct shim_vma * tmp,
-                                    const void * addr, uint64_t length)
+/*
+ * Return true if [s, e) is exactly the area represented by vma.
+ */
+static inline bool test_vma_equal (struct shim_vma * vma,
+                                   void * s, void * e)
 {
-    return tmp->addr <= addr &&
-           tmp->addr + tmp->length >= addr + length;
+    return vma->start == s && vma->end == e;
 }
 
-static inline bool test_vma_startin (struct shim_vma * tmp,
-                                    const void * addr, uint64_t length)
+/*
+ * Return true if [s, e) is part of the area represented by vma.
+ */
+static inline bool test_vma_contain (struct shim_vma * vma,
+                                     void * s, void * e)
 {
-    return tmp->addr >= addr &&
-           tmp->addr < addr + length;
+    return vma->start <= s && vma->end >= e;
 }
 
-static inline bool test_vma_endin (struct shim_vma * tmp,
-                                  const void * addr, uint64_t length)
+/*
+ * Return true if [s, e) contains the starting address of vma.
+ */
+static inline bool test_vma_startin (struct shim_vma * vma,
+                                     void * s, void * e)
 {
-    return tmp->addr + tmp->length > addr &&
-           tmp->addr + tmp->length <= addr + length;
+    return vma->start >= s && vma->start < e;
 }
 
-static inline bool test_vma_overlap (struct shim_vma * tmp,
-                                    const void * addr, uint64_t length)
+/*
+ * Return true if [s, e) contains the ending address of vma.
+ */
+static inline bool test_vma_endin (struct shim_vma * vma,
+                                   void * s, void * e)
 {
-    return test_vma_contain(tmp, addr, 1) ||
-           test_vma_contain(tmp, addr + length - 1, 1) ||
-           test_vma_startin(tmp, addr, length);
+    return vma->end > s && vma->end <= e;
```

Comment:
for [s. e) range should we check vma->end >=s and vma->end < e ? 
Same as above ?

Reply:
Hmm, I think the implementation is correct, but the comment is a bit misleading.
@chiache: Wasn't this function supposed to test whether the last byte of `vma` (not the end address) is contained in `[s, e)`?